### PR TITLE
Actions tester Server

### PR DESCRIPTION
**Pull request by dlasky**
Adds support for websocket server + manifest route needed by the actions tester UI

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [ ] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [Segmenters] Tested in the staging environment

### CODE REVIEW

File: packages/cli/src/commands/serve.ts [hunk 1]
```diff
@@ -124,6 +136,8 @@ export default class Serve extends Command {
 
     watcher.once('ready', () => {
       this.log(chalk.greenBright`Watching required files for changes .. `)
+      this.log(chalk.greenBright`Visit https://app.segment.com/dev_center/actions-tester to preview your integration.`)
+      void open('https://app.segment.com/dev_center/actions-tester')
```
Comment (mericsson) [inline packages/cli/src/commands/serve.ts:140]: Consider this format for consistency

```suggestion
      this.log(chalk.greenBright`Visit https://app.segment.com/dev-center/actions-tester to preview your integration.`)
      void open('https://app.segment.com/dev-center/actions-tester')
```

File: packages/cli/src/commands/serve.ts [hunk 2]
```diff
@@ -124,6 +136,8 @@ export default class Serve extends Command {
 
     watcher.once('ready', () => {
       this.log(chalk.greenBright`Watching required files for changes .. `)
+      this.log(chalk.greenBright`Visit https://app.segment.com/dev-center/actions-tester to preview your integration.`)
+      void open('https://app.segment.com/dev-center/actions-tester')
```
Comment (nick-Ag) [inline packages/cli/src/commands/serve.ts:140]: Will this link be opened every time someone runs `./bin/run serve`?

Comment (dlasky) [reply packages/cli/src/commands/serve.ts:140]: yes, similar to starting `app`

File: packages/cli/src/lib/server.ts [hunk 3]
```diff
@@ -13,15 +14,53 @@ import {
...
+const marshalError = (err: Error): ErrorOutput => {
+  const error = err as ResponseError
+  let statusCode = error?.status ?? 500
+  let msg = error?.message
+  let fields: { [key: string]: string } | undefined
+
+  if (error.name === 'AggregateAjvError') {
+    fields = {}
+    const ajvErr = error as AggregateAjvError
+    for (const fieldError of ajvErr) {
```
Comment (gurdasnijor) [inline packages/cli/src/lib/server.ts:39]: One of the downsides of working with ajv in a ts codebase :/ 

I'd much prefer something like https://github.com/jquense/yup#getting-started (but changing our json validation lib is a bit out of scope for this i'd imagine :) )

File: packages/cli/src/commands/serve.ts [hunk 4]
```diff
@@ -77,6 +79,16 @@ export default class Serve extends Command {
       cwd: process.cwd()
     })
 
+    const DEFAULT_PORT = 3000
+    const port = parseInt(process.env.PORT ?? '', 10) || DEFAULT_PORT
+    const wss = new WebSocketServer({ port: port + 1 })
```
Comment (nick-Ag) [inline packages/cli/src/commands/serve.ts:84]: Will the existing server at port 3000 and the new web socket always be running together? I wonder if we should allow running one or the other exclusively? (Not sure if there's a noticeable performance diff)

File: packages/cli/src/lib/server.ts (hunk 3 above)
Comment (dlasky) [reply packages/cli/src/lib/server.ts:39]: that was just a refactor of some existing stuff--im not quite sure what it was being used for as I couldn't get it to reproduce, but I wanted to bring a consistent error handler on this side so the client can parse it uniformly

File: packages/cli/src/commands/serve.ts (hunk 4 above)
Comment (dlasky) [reply packages/cli/src/commands/serve.ts:84]: they are currently both required to function, the websocket server is used only for telling the client to update.

File: packages/cli/src/lib/server.ts (hunk 3 above)
Comment (gurdasnijor) [reply packages/cli/src/lib/server.ts:39]: 👍🏽

File: packages/cli/src/lib/server.ts [hunk 5]
```diff
@@ -13,15 +14,53 @@ import {
...
+  }
+
+  return { statusCode, message: msg, stack: err.stack?.split('\n'), fields, requestError: true }
+}
+
 const app = express()
 app.use(express.json())
+app.use(
+  cors({
+    origin: ['https://app.segment.build', 'https://app.segment.com', 'http://localhost:8000']
```
Comment (gurdasnijor) [inline packages/cli/src/lib/server.ts:58]: Should we get our eu subdomains in here too?

Comment (dlasky) [reply packages/cli/src/lib/server.ts:58]: Good question, its technically not user data so i'm not sure it matters for GDPR? but i suppose if a destination author's workspace is EU they would be at that domain?

File: packages/cli/src/lib/server.ts [hunk 6]
```diff
@@ -103,6 +142,13 @@ function setupRoutes(def: DestinationDefinition | null): void {
 
   const router = express.Router()
 
+  router.get(
+    '/manifest',
+    asyncHandler(async (_, res: express.Response) => {
+      res.json(destination.definition)
```
Comment (gurdasnijor) [inline packages/cli/src/lib/server.ts:148]: I've been out of express-land for quite some time - is it not necessary to call `next()` to yield control to the next middleware in the stack anymore?

Comment (dlasky) [reply packages/cli/src/lib/server.ts:148]: i think thats only for middleware, routes just get req, res cause they are effectively the 'last' step in the chain

**Review by gurdasnijor — APPROVED** [review]
Still bringing up a local env to do some functional testing of this + https://github.com/segmentio/app/pull/12997 - going to give the 👍🏽 for now to unblock getting this released and will share results/feedback after - awesome work!  👏🏽

File: packages/cli/src/lib/server.ts (hunk 6 above)
Comment (gurdasnijor) [reply packages/cli/src/lib/server.ts:148]: Ahhhhh missed that - thanks!

File: packages/cli/src/commands/serve.ts (hunk 2 above)
Comment (dlasky) [reply packages/cli/src/commands/serve.ts:140]: i guess it does make sense to comment this for now until the UI portion is accessible to all users

**Pull request merged**